native_client/beam_search.cc

```diff
@@ -0,0 +1,546 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This test illustrates how to make use of the CTCBeamSearchDecoder using a
+// custom BeamScorer and BeamState based on a dictionary with a few artificial
+// words.
+#include "tensorflow/core/util/ctc/ctc_beam_search.h"
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/kernels/bounds_check.h"
+
+#include "kenlm/lm/model.hh"
+
+#include "alphabet.h"
+#include "trie_node.h"
+
+namespace tf = tensorflow;
+using tf::shape_inference::DimensionHandle;
+using tf::shape_inference::InferenceContext;
+using tf::shape_inference::ShapeHandle;
+
+REGISTER_OP("CTCBeamSearchDecoderWithLM")
+    .Input("inputs: float")
+    .Input("sequence_length: int32")
+    .Attr("model_path: string")
+    .Attr("trie_path: string")
+    .Attr("alphabet_path: string")
+    .Attr("beam_width: int >= 1 = 100")
+    .Attr("top_paths: int >= 1 = 1")
+    .Attr("merge_repeated: bool = true")
+    .Output("decoded_indices: top_paths * int64")
+    .Output("decoded_values: top_paths * int64")
+    .Output("decoded_shape: top_paths * int64")
+    .Output("log_probability: float")
+    .SetShapeFn([](InferenceContext *c) {
+      ShapeHandle inputs;
+      ShapeHandle sequence_length;
+
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 3, &inputs));
+      TF_RETURN_IF_ERROR(c->WithRank(c->input(1), 1, &sequence_length));
+
+      // Get batch size from inputs and sequence_length.
+      DimensionHandle batch_size;
+      TF_RETURN_IF_ERROR(
+          c->Merge(c->Dim(inputs, 1), c->Dim(sequence_length, 0), &batch_size));
+
+      tf::int32 top_paths;
+      TF_RETURN_IF_ERROR(c->GetAttr("top_paths", &top_paths));
+
+      // Outputs.
+      int out_idx = 0;
+      for (int i = 0; i < top_paths; ++i) {  // decoded_indices
+        c->set_output(out_idx++, c->Matrix(InferenceContext::kUnknownDim, 2));
+      }
+      for (int i = 0; i < top_paths; ++i) {  // decoded_values
+        c->set_output(out_idx++, c->Vector(InferenceContext::kUnknownDim));
+      }
+      ShapeHandle shape_v = c->Vector(2);
+      for (int i = 0; i < top_paths; ++i) {  // decoded_shape
+        c->set_output(out_idx++, shape_v);
+      }
+      c->set_output(out_idx++, c->Matrix(batch_size, top_paths));
+      return tf::Status::OK();
+    })
+    .Doc(R"doc(
```

Comment:
I think the doc needs to be updated to include the new parameters

Reply:
I didn't change this line so GitHub is still showing the comment but it's been fixed.